### PR TITLE
feat(inverted_index.create): add read/write for external intermediate files

### PR DESCRIPTION
**Pull request by zhongzc**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

* Add related format and codec for intermediate files
* Add the facility for reading and writing intermediate files, using `FramedRead` and `FramedWrite`, to achieve streaming read-write capabilities

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
#2705

next PRs to be proposed:
[feat(inverted_index.create): add external sorter](https://github.com/zhongzc/greptimedb/pull/3)
[feat(inverted_index.create): add index creator](https://github.com/zhongzc/greptimedb/pull/4)

### CODE REVIEW

**Review by waynexia — APPROVED** [review]
LGTM :yum:

File: src/index/src/inverted_index/create/sort/intermediate_rw.rs [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+            .write_all(codec_magic)
+            .await
+            .context(WriteSnafu)?;
+
+        let value_stream = stream::iter(values.into_iter().map(Ok));
+        let frame_write = FramedWrite::new(&mut self.writer, encoder);
+        value_stream.forward(frame_write).await?;
+
+        self.writer.flush().await.context(FlushSnafu)?;
+        self.writer.close().await.context(CloseSnafu)
```
Comment (fengjiachun) [inline src/index/src/inverted_index/create/sort/intermediate_rw.rs:55]: If `flush` fails, do we need also call `close`?

File: src/index/Cargo.toml [hunk 2]
```diff
@@ -7,11 +7,13 @@ license.workspace = true
 [dependencies]
 async-trait.workspace = true
 bytemuck.workspace = true
+bytes = "0.5.6"
```
[the remark this answers is not in the document]
Comment (zhongzc) [reply src/index/Cargo.toml:10]: https://github.com/GreptimeTeam/greptimedb/blob/f592c18766464e2575a0c9f01fd62e846a3fdf80/Cargo.lock#L3516-L3526

[the remark this answers is not in the document]
Comment (zhongzc) [reply src/index/Cargo.toml:10]: I can't operate the `Bytes` while using a newer version.

File: src/index/src/inverted_index/create/sort/intermediate_rw.rs (hunk 1 above)
Comment (zhongzc) [reply src/index/src/inverted_index/create/sort/intermediate_rw.rs:55]: I think the `Drop` impl of the most underlying writer should do the `close` in such a situation.

**Review by evenyag — APPROVED** [review]
LGTM

File: src/index/Cargo.toml (hunk 2 above)
Comment (killme2008) [inline src/index/Cargo.toml:10]:
```suggestion
bytes = "0.5"
```

[the remark this answers is not in the document]
Comment (evenyag) [reply src/index/Cargo.toml:10]: Can we downgrade the bytes crate we use in the workspace?

[the remark this answers is not in the document]
Comment (evenyag) [reply src/index/Cargo.toml:10]: Oh, I see. The last update of `futures_codec` was 3 years ago...
https://github.com/matthunz/futures-codec

File: src/index/src/inverted_index/create/sort/intermediate_rw/codec_v1.rs [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        // decode value len
+        if src.len() < U64_LENGTH {
+            return Ok(None);
+        }
+        let (value_len, buf) = src.split_at(U64_LENGTH);
+        let value_len = u64::from_le_bytes(value_len.try_into().unwrap()) as usize;
+
+        // decode value
+        if buf.len() < value_len {
+            return Ok(None);
```
Comment (killme2008) [inline src/index/src/inverted_index/create/sort/intermediate_rw/codec_v1.rs:80]: Is such a case an error or normal? It's possible when writing fails, but I am not sure if it's ok to return `None`.

**Review by fengjiachun — APPROVED** [review]
LGTM

File: src/index/Cargo.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (zhongzc) [reply src/index/Cargo.toml:10]: Very good investigation, I like the implementation of `futures_codec`, the `bytes` version was a hindrance, things seem to be getting better now.

Comment (zhongzc) [reply src/index/Cargo.toml]: Updated to "0.1"

File: src/index/src/inverted_index/create/sort/intermediate_rw/codec_v1.rs (hunk 3 above)
Comment (zhongzc) [reply src/index/src/inverted_index/create/sort/intermediate_rw/codec_v1.rs:80]: It's a framed read. `FramedReader` may read up to the end of a frame (not enough remaining bytes for one Item), in which case `FramedReader` needs to continue adding bytes and retry. Only when there are no subsequent bytes will `FramedReader` raise an error.

Comment (zhongzc) [reply src/index/src/inverted_index/create/sort/intermediate_rw/codec_v1.rs:80]: When writing fails, it eventually returns an error thrown by the reader.

**Pull request merged**